File: prototype_python/game_state.py
"""Top-level game-state model for the portable rules engine."""

from typing import Optional

from .constants import (
    DAY_NIGHT_NONE,
    DEFAULT_COMMANDER_DAMAGE_THRESHOLD,
    DEFAULT_PLAYERS,
    DEFAULT_POISON_THRESHOLD,
    DEFAULT_STARTING_LIFE,
    MAX_COMMANDERS_PER_PLAYER,
    MAX_PLAYERS,
    MIN_PLAYERS,
)
from .player import Player


class GameState:
    """All state for one Commander game.

    Player IDs are zero-based indexes. The commander damage matrix is indexed as
    ``[target_player][source_player][commander_slot]``.
    """

    def __init__(
        self,
        player_count: int = DEFAULT_PLAYERS,
        starting_life: int = DEFAULT_STARTING_LIFE,
        poison_threshold: int = DEFAULT_POISON_THRESHOLD,
        commander_damage_threshold: int = DEFAULT_COMMANDER_DAMAGE_THRESHOLD,
        partner_mode_enabled: bool = False,
        confirm_eliminations: bool = True,
    ) -> None:
        if player_count < MIN_PLAYERS or player_count > MAX_PLAYERS:
            raise ValueError(
                f"player_count must be between {MIN_PLAYERS} and {MAX_PLAYERS}"
            )
        if poison_threshold <= 0:
            raise ValueError("poison_threshold must be positive")
        if commander_damage_threshold <= 0:
            raise ValueError("commander_damage_threshold must be positive")

        self.player_count = player_count
        self.starting_life = starting_life
        self.players = [
            Player(player_id, f"P{player_id + 1}", starting_life)
            for player_id in range(player_count)
        ]
        self.poison_threshold = poison_threshold
        self.commander_damage_threshold = commander_damage_threshold
        self.partner_mode_enabled = partner_mode_enabled
        self.commander_damage = self._new_commander_damage_matrix()
        self.turn_number = 1
        self.active_player: Optional[int] = 0
        self.monarch_player: Optional[int] = None
        self.initiative_player: Optional[int] = None
        self.storm_count = 0
        self.day_night_state = DAY_NIGHT_NONE
        self.confirm_eliminations = confirm_eliminations

    def _new_commander_damage_matrix(self) -> list[list[list[int]]]:
        return [
            [
                [0] * MAX_COMMANDERS_PER_PLAYER
                for _source in range(self.player_count)
            ]
            for _target in range(self.player_count)
        ]

    def get_player(self, player_id: int) -> Player:
        if player_id < 0 or player_id >= self.player_count:
            raise ValueError(f"invalid player_id: {player_id}")
        return self.players[player_id]

    def validate_commander_slot(self, commander_slot: int) -> None:
        if commander_slot < 0 or commander_slot >= MAX_COMMANDERS_PER_PLAYER:
            raise ValueError(f"invalid commander_slot: {commander_slot}")

    def reset_commander_damage(self) -> None:
        self.commander_damage = self._new_commander_damage_matrix()

File: gbc_src/game_state.c
#include "game_state.h"

static void initialize_player_names(GameState *game) {
    uint8_t player_id;
    char name[3];

    name[0] = 'P';
    name[2] = '\0';
    for (player_id = 0u; player_id < game->player_count; player_id++) {
        name[1] = (char)('1' + player_id);
        player_init(
            &game->players[player_id],
            player_id,
            name,
            game->starting_life
        );
    }
}

static void clear_commander_damage(GameState *game) {
    uint8_t target;
    uint8_t source;
    uint8_t commander_slot;

    for (target = 0u; target < MAX_PLAYERS; target++) {
        for (source = 0u; source < MAX_PLAYERS; source++) {
            for (
                commander_slot = 0u;
                commander_slot < MAX_COMMANDERS_PER_PLAYER;
                commander_slot++
            ) {
                game->commander_damage[target][source][commander_slot] = 0u;
            }
        }
    }
}

void game_state_reset(GameState *game) {
    uint8_t player_id;

    for (player_id = 0u; player_id < game->player_count; player_id++) {
        player_reset(&game->players[player_id], game->starting_life);
    }
    clear_commander_damage(game);
    game->turn_number = 1u;
    game->active_player = 0;
    game->monarch_player = NO_PLAYER;
    game->initiative_player = NO_PLAYER;
    game->storm_count = 0u;
    game->day_night_state = 0u;
}

void game_state_init(
    GameState *game,
    uint8_t player_count,
    int16_t starting_life
) {
    if (player_count < MIN_PLAYERS) {
        player_count = MIN_PLAYERS;
    } else if (player_count > MAX_PLAYERS) {
        player_count = MAX_PLAYERS;
    }

    game->player_count = player_count;
    game->starting_life = starting_life;
    game->poison_threshold = DEFAULT_POISON_THRESHOLD;
    game->commander_damage_threshold = DEFAULT_COMMANDER_DAMAGE_THRESHOLD;
    game->partner_mode_enabled = 0u;
    game->confirm_eliminations = 1u;

    initialize_player_names(game);
    game_state_reset(game);
}
